File: StrategyPattern/strategyPattern.java
public interface FlyBehaviour{
    void fly();
}

public class FlyWithWings implements FlyBehaviour{
    public void fly(){
        System.out.println("I'm flying!");
    }
}

public class FlyNoWay implements FlyBehaviour{
    public void fly(){
        System.out.println("I can't fly!");
    }
}

public interface QuackBehaviour{
    void quack();
}

public class Quack implements QuackBehaviour{
    public void quack(){
        System.out.println("I can quack.");
    }
}

public class Squeak implements QuackBehaviour{
    public void quack(){
        System.out.println("I can squeak.");
    }
}

public class Mute implements QuackBehaviour{
    public void quack(){
        System.out.println("I am a silent duck.");
    }
}

public abstract class Duck{
    FlyBehaviour flyBehaviour;
    QuackBehaviour quackBehaviour;

    abstract void display();
    public void performFly(){
        flyBehaviour.fly();
    }

    public void performQuack(){
        quackBehaviour.quack();
    }

    public void swim (){
        System.out.println("All ducks can swim.So I can too.");
    }
}

public class MallardDuck extends Duck{

    public MallardDuck(){
        quackBehaviour = new Quack();
        flyBehaviour = new FlyWithWings();
    }

    public void display(){
        System.out.println("I'm a  Mallard Duck.")
    }

}

public class MiniDuckSimulator{
    public static void main(String[] args){
        Duck mallard = new MallardDuck();
        mallard.performFly();
        mallard.performQuack();
        mallard.display();
    }
}

File: StrategyPattern/strategy_pattern.py
class fly_behaviour:
    def fly(self):
        print("")


class fly_with_wings(fly_behaviour):
    def fly(self):
        print("I can fly")


class fly_noway(fly_behaviour):
    def fly(self):
        print("I can't fly")


class quack_behaviour:
    def sounds(self):
        print("")


class quack(quack_behaviour):
    def sounds(self):
        print("I quack!")


class squeak(quack_behaviour):
    def sounds(self):
        print("I squeak!")


class mute(quack_behaviour):
    def sounds(self):
        print("I am shy!")


class duck:
    def __init__(self):
        self.flying = fly_behaviour()
        self.quacking = quack_behaviour()
        self.mytype = "I am a normal duck."

    def perform_swim(self):
        print("I can swim!")

    def __str__(self):
        self.flying.fly()
        self.quacking.sounds()
        return self.mytype


class mallard_duck(duck):
    def __init__(self):
        super()
        self.flying = fly_with_wings()
        self.quacking = quack()
        self.mytype = "I am a mallard duck."


class rubber_duck(duck):
    def __init__(self):
        super()
        self.flying = fly_noway()
        self.quacking = mute()
        self.mytype = "I am a rubber duck."


mallrd = mallard_duck()
print(mallrd)

toy = rubber_duck()
print(toy)
